sae_lens/toolkit/sae_patcher.py

```diff
@@ -0,0 +1,135 @@
+from typing import Literal, Protocol
+
+import torch
+from jaxtyping import Float
+from transformer_lens.hook_points import HookPoint
+
+from sae_lens import SparseAutoencoder
+
+
+class TransformerLensForwardHook(Protocol):
+    def __call__(self, orig: torch.Tensor, hook: HookPoint) -> torch.Tensor:
+        raise NotImplementedError
+
+
+class TransformerLensBackwardHook(Protocol):
+    def __call__(self, orig: torch.Tensor, hook: HookPoint) -> tuple[torch.Tensor]:
+        raise NotImplementedError
+
+
+ForwardHookData = tuple[str, TransformerLensForwardHook]
+BackwardHookData = tuple[str, TransformerLensBackwardHook]
+
+NodeType = Literal["feature", "error", "all"]
+
+
+class SAEPatcher:
+    """Patches an SAE into the computational graph of a HookedTransformer
+
+    Usage:
+
+    sae_patcher = SAEPatcher(sae)
+    with model.hooks(
+        fwd_hooks=[sae_patcher.get_forward_hook()]
+        bwd_hooks=[sae_patcher.get_backward_hook()]
+    ):
+        ...
+    """
+
+    sae: SparseAutoencoder
+    sae_feature_acts: Float[torch.Tensor, "n_batch n_token d_sae"]
+    sae_errors: Float[torch.Tensor, "n_batch n_token d_model"]
+
+    def __init__(self, sae: SparseAutoencoder):
+        self.sae = sae
+        self.sae_feature_acts = torch.tensor([])
+        self.sae_errors = torch.tensor([])
+
+    def _forward_hook_fn(
+        self,
+        orig: Float[torch.Tensor, "n_batch n_token d_model"],
+        hook: HookPoint,
+    ) -> Float[torch.Tensor, "n_batch n_token d_model"]:
+        """Forward hook to patch the SAE into the computational graph
+
+        orig: Original activations
+        """
+
+        a_orig = orig
+        z_sae = self.sae.encode(a_orig)
+        # keep pyright happy
+        assert isinstance(z_sae, torch.Tensor)
+        a_sae = self.sae.decode(z_sae)
+        a_err = a_orig - a_sae.detach()
+
+        # Track the gradients
+        assert z_sae.requires_grad
+        z_sae.retain_grad()
+        assert a_err.requires_grad
+        a_err.retain_grad()
+
+        # Store values for later use
+        self.sae_feature_acts = z_sae
+        self.sae_errors = a_err
+
+        a_rec = a_sae + a_err
+        return a_rec
+
+    @property
+    def sae_nodes(self) -> Float[torch.Tensor, "n_batch n_token (d_sae + d_model)"]:
+        return torch.cat([self.sae_feature_acts, self.sae_errors], dim=-1)
+
+    def _backward_hook_fn(
+        self,
+        orig: Float[torch.Tensor, "n_batch n_token d_model"],
+        hook: HookPoint,
+    ) -> tuple[Float[torch.Tensor, "n_batch n_token d_model"]]:
```

Comment:
Why is this function needed? It looks like it's just the identity

Reply:
The idea here is that, when we patch the model with an SAE at an activation, there are actually two copies of the activation we care about: $a^u$, the copy which is fed into the SAE, and $a^d$, the copy which comes out of it. 

The values of $a^u$ and $a^d$ are the same, but the specific implementation means that gradients from $a^d$ do not get passed to $a^u$. (Concretely, we called `detach()` somewhere in there).

To fix this, we need a backward hook which directly passes the gradient of $a^d$ to $a^u$. This hooks does that, since the input is the gradient of $a^d$ and the output is the new gradient of $a^u$.

Reply:
See Marks' paper, appendix A3, 'Stop gradients` and `Passthrough gradients` for more explanation

Reply:
Yeah, basically it boils down to the fact that if we don't have this then everything before that layer won't receive gradient!